# Lab work for STATS: Advanced Statistical Methodlogy


http://www.stat.purdue.edu/~tlzhang/stat526/homework.html

In [1]:
import pandas as pd
import numpy as np

# import plotting libraries
import matplotlib.pyplot as plt
import seaborn as sns

%matplotlib inline
plt.style.use('seaborn-white')

# Statistical Packages
import scipy.stats as stats
import statsmodels.api as sm
import statsmodels.formula.api as smf

## LAB 2

> following table was reported
* a. Test independence by two sample binomial method and odds ratio respectively
* b. Derive a 95% confidence interval for the odds ratio
* c. Interpret odds ratio results

http://www.stat.purdue.edu/~tlzhang/stat526/hw2_526.pdf

In [5]:
gun = pd.DataFrame(columns=['favor', "oppose"], index=['favor', 'oppose'])
gun.loc['favor'] = [784, 236]
gun.loc['oppose'] = [311, 66]
#gun_t = sm.stats.Table(gun)
tab = sm.stats.Table2x2(gun.values)
print(tab.table_orig)

chi2, p, dof, ex = stats.chi2_contingency(gun, correction=False)

print("\nVictim = Black:")
print("Pearson's Chi-squared test with Yates' continuity correction")
print("X-squared: ", round(chi2, 3))
print("p-value  : ", round(p, 3))
print("degree of f: ", dof)

hi, low = tab.oddsratio_confint()
print("Odds Ratio CI: ", (round(hi, 3), round(low, 3)))
print("\n2x2 Summary:")
print(tab.summary())

[[784 236]
 [311  66]]

Victim = Black:
Pearson's Chi-squared test with Yates' continuity correction
X-squared:  5.15
p-value  :  0.023
degree of f:  1
Odds Ratio CI:  (0.521, 0.954)

2x2 Summary:
               Estimate   SE   LCB    UCB   p-value
---------------------------------------------------
Odds ratio        0.705        0.521  0.954   0.024
Log odds ratio   -0.350 0.155 -0.652 -0.047   0.024
Risk ratio        0.932        0.880  0.987   0.016
Log risk ratio   -0.071 0.029 -0.128 -0.013   0.016
---------------------------------------------------


1. __Test of independence__, with a $\chi^2$ p-value = 0.023, this is less than 0.05, providing significant evidence to reject the null hypothesis that a relationship exist between `gun_registration` and the `death+penalty`.
2. __Odds Ratio:__, with $\hat\theta=0.705$ on the odds ratio, we can conclude that favors in gun registration is about 71% least likely than than those who favor the death penalty.
3. __Odds Ratio Interpretation:__ The 95% confidence interval for the __odds ratio__ of those that favor gun registration in comparision to opposing gun resitration is $\hat\theta=0.705$, which is between (0.521 and 0.954). Since the interval is not between 1, indicating this the relationship is similar and not significant (FTR). 


__People who favor Gun Registration
faver less on Death Penalty. Of those who favor Gun Registration, they were 30% less favor the Death 
Penalty than those who oppose Gun Registration.__

In [36]:
result = [577,34,682,57,164,4,245,74]
d = {
    'age': np.repeat(('young','young','young','young',
                     'old','old','old','old',), (result)),
    'smoking' : np.repeat(('nonsmoker', 'nonsmoker', 'smoker', 'smoker',
                          'nonsmoker', 'nonsmoker', 'smoker', 'smoker',), (result)),
    'breathing': np.repeat(('normal', 'not_normal', 'normal', 'not_normal',
                           'normal', 'not_normal', 'normal', 'not_normal',), (result))
}

breath = pd.DataFrame(d, columns=['age', 'smoking','breathing'])
breath_tb = pd.crosstab([breath.age, breath.smoking], breath.breathing)
breath_tb

breathing        normal  not_normal
age   smoking                      
old   nonsmoker     164           4
      smoker        245          74
young nonsmoker     577          34
      smoker        682          57

In [37]:
def odds_ratio_test(table):
    print(table, "\n")

    chi2, p, dof, ex = stats.chi2_contingency(table, correction=False)

    print("Pearson's Chi-squared test with Yates' continuity correction")
    print("X-squared: ", chi2)
    print("p-value  : ", p)
    print("degree of f: ", dof)

    oddsratio, pvalue = stats.fisher_exact(table, alternative='two-sided')
    print("Odds Ratio: ", round(oddsratio, 3))

    tab = sm.stats.Table2x2(table.values)
    print("Odds Ratio CI: ", tab.oddsratio_confint())

    print("\n2x2 Summary \n:", tab.summary())

> estimate the odds ratio, give 95% CI and test indedpendece for age less than 40 and age between 40 adn 59, respevtively.

In [38]:
young =pd.crosstab([breath[breath.age=="young"].smoking], breath.breathing)
odds_ratio_test(young)

breathing  normal  not_normal
smoking                      
nonsmoker     577          34
smoker        682          57 

Pearson's Chi-squared test with Yates' continuity correction
X-squared:  2.455918816761549
p-value  :  0.11708263954444463
degree of f:  1
Odds Ratio:  1.418
Odds Ratio CI:  (0.914446140736943, 2.1999693125303468)

2x2 Summary 
:                Estimate   SE   LCB    UCB  p-value
--------------------------------------------------
Odds ratio        1.418        0.914 2.200   0.119
Log odds ratio    0.350 0.224 -0.089 0.788   0.119
Risk ratio        1.023        0.995 1.053   0.112
Log risk ratio    0.023 0.014 -0.005 0.051   0.112
--------------------------------------------------


__interpretation__ 
* __Independence:__ with a $\chi^2=2.455$ and the p-value = 0.117, there is overwelming evidence that the smokering for ages < 40 are __NOT independent, in fact they are dependent (related).__ We probabily can say that smoking will cause abnormal breaking test results in a long period of time. 
* __Odds Ratio:__ we can conclude with a $\hat\theta=1.418$, the odds of non-smoker having normal breaking retest results are 42% higher as they are for smokers, given a young participant. We are 95% confidence the true population based on the odds ratio is between 0.883 and 2.124.

In [39]:
old =pd.crosstab([breath[breath.age=="old"].smoking], breath.breathing)
odds_ratio_test(old)

breathing  normal  not_normal
smoking                      
nonsmoker     164           4
smoker        245          74 

Pearson's Chi-squared test with Yates' continuity correction
X-squared:  35.45096789317344
p-value  :  2.6155238327504702e-09
degree of f:  1
Odds Ratio:  12.384
Odds Ratio CI:  (4.441584053896396, 34.52717920803201)

2x2 Summary 
:                Estimate   SE   LCB   UCB   p-value
--------------------------------------------------
Odds ratio       12.384       4.442 34.527   0.000
Log odds ratio    2.516 0.523 1.491  3.542   0.000
Risk ratio        1.271       1.191  1.356   0.000
Log risk ratio    0.240 0.033 0.175  0.305   0.000
--------------------------------------------------


__interpretation__ 
* __Independence:__ with a $\chi^2=35.45$ and the p-value = <0.001, there is overwelming evidence that the smokering for ages > 40 are independent. 
* __Odds Ratio:__ we can conclude with a $\hat\theta=12.384$, the odds of non-smoker having normal breaking retest results are __over 12 time higher__ as they are for smokers, given a young participant. We are 95% confidence the true population based on the odds ratio is between 4.442 and 34.527.

In [40]:
both =pd.crosstab([breath.smoking], breath.breathing)
odds_ratio_test(both)

breathing  normal  not_normal
smoking                      
nonsmoker     741          38
smoker        927         131 

Pearson's Chi-squared test with Yates' continuity correction
X-squared:  30.242136392093776
p-value  :  3.813341838310365e-08
degree of f:  1
Odds Ratio:  2.756
Odds Ratio CI:  (1.896169365455782, 4.004748246701448)

2x2 Summary 
:                Estimate   SE   LCB   UCB  p-value
-------------------------------------------------
Odds ratio        2.756       1.896 4.005   0.000
Log odds ratio    1.014 0.191 0.640 1.387   0.000
Risk ratio        1.086       1.056 1.116   0.000
Log risk ratio    0.082 0.014 0.054 0.110   0.000
-------------------------------------------------


__interpretation__ 
* __Independence:__ with a $\chi^2=28.49$ and the p-value = <0.001, there is overwelming evidence that the smokering are independent from non-smokers. 
* __Odds Ratio:__ we can conclude with a $\hat\theta=2.681$, the odds of non-smoker having normal breathing test results are __2.7 time higher__ as they are for smokers, given a young participant. We are 95% confidence the true population based on the odds ratio is between 1.845 and 3.897.

In [35]:
9.386105969429338e-08

9.386105969429338e-08